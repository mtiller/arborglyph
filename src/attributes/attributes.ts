import { TreeMap } from "../maps";

/**
 * An attribute is simply a function that we can call passing in the node
 * identifier and getting back some value.  The details of how the value
 * is computed (lazy, caching, etc) are hidden.
 */
export interface Attribute<R> {
  (nid: string): R;
  // TODO: Get rid of this
  invalidate(): void;
}

/**
 * The `AttributeTypes` type is simply a mapping of attribute name
 * to the type of that attribute.
 */
export type AttributeTypes = Record<string, any>;

/**
 * Given an `AttributesTypes` type, formulate the type of the
 * `attrs` argument containing previously defined attributes
 * when evaluating a new attribute.
 */
export type DefinedAttributes<A extends AttributeTypes> = {
  [P in keyof A]: Attribute<A[P]>;
};

export type ExtendedBy<
  A extends AttributeTypes,
  N extends string,
  R
> = DefinedAttributes<A & Record<N, R>>;

export type AttributeConstructor<
  N extends string,
  T,
  D extends AttributeTypes, // What this function depends on
  R
> = <A extends D>(
  tree: TreeMap<T>,
  base: DefinedAttributes<D>,
  ext: DefinedAttributes<A>
) => ExtendedBy<A, N, R>;
