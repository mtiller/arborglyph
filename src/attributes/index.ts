export * from "./depth";
export * from "./descendents";
export * from "./name-map";
export * from "./path";
