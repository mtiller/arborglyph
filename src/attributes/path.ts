import {
  inherited,
  InheritedAttributeDefinition,
  InheritedOptions,
} from "../kinds/definitions";

export function evalPath<T>(
  opts?: Partial<InheritedOptions>
): InheritedAttributeDefinition<T, string> {
  return inherited(
    "path",
    ({ node, parent }) => {
      return parent.caseOf({
        Nothing: () => "root",
        Just: (p) => {
          if (node === p.node) throw new Error("Node was its own parent?!?");
          for (const [key, value] of Object.entries(p.node)) {
            if (value === node) {
              return `${p.attr}/${key}`;
            }
          }
          throw new Error(
            "Unable to find node in parent (this shouldn't happen)"
          );
        },
      });
    },
    opts
  );
}
