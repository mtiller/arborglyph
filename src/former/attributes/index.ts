export * from "./arborglyph";
export * from "./attributes";
export * from "./derived";
export * from "./inherited";
export * from "./memoize";
export * from "./synthetic";
