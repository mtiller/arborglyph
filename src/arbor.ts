import { DerivedEvaluator } from "./kinds/derived";
import {
  InheritedAttributeEvaluator,
  InheritedOptions,
  reifyInheritedAttribute,
} from "./kinds/inherited";
import {
  reifySyntheticAttribute,
  SyntheticAttributeEvaluator,
  SyntheticOptions,
} from "./kinds/synthetic";
import { Attribute } from "./kinds/attributes";
import { AttributeDefinition } from "./kinds/definitions";
import { assertUnreachable } from "./utils";

/**
 * This file contains a couple different ways to represent a tree.  It is
 * worth noting that these are all "top down".  This is really the only
 * practical representation because a bottom up tree isn't really traversable
 * because you'd have to know all the leaves in order to ensure that you could
 * visit every node.
 */

export type IndexedChildren<T> = (x: T) => T[];
export type NamedChildren<T> = (x: T) => Record<string, T>;
export type ListChildren<T> = IndexedChildren<T> | NamedChildren<T>;

export interface ArborOptions<T> {
  // wrappers
  // inh (options, no R)
  // syn (options, no R)
}

/** A potentially convenient class, not sure what I think about it yet. */
export class Arbor<T extends object> {
  protected reified: Map<AttributeDefinition<any, any>, Attribute<any, any>>;
  constructor(
    public root: T,
    public list: ListChildren<T>,
    opts: ArborOptions<T> = {}
  ) {
    this.reified = new Map();
  }
  add<R>(def: AttributeDefinition<T, R>): Attribute<T, R> {
    if (this.reified.has(def)) {
      return this.reified.get(def) as Attribute<T, R>;
    }
    switch (def.type) {
      case "syn": {
        const r = reifySyntheticAttribute<T, R>(
          this.root,
          this.list,
          def.f,
          {}
        );
        this.reified.set(def, r);
        return r;
      }
      case "inh": {
        const r = reifyInheritedAttribute<T, R>(
          this.root,
          this.list,
          def.f,
          {}
        );
        this.reified.set(def, r);
        return r;
      }
      case "der": {
        // TODO: As attribute gets expanded, more will probably be needed here.
        const ret = (x: T) => def.f(x);
        this.reified.set(def, ret);
        return ret;
      }
      case "trans": {
        const attr = this.add(def.attr);
        const ret = (x: T) => def.f(attr(x));
        this.reified.set(def, ret);
        return ret;
      }
    }
    return assertUnreachable(def);
  }
  inh<R>(
    f: InheritedAttributeEvaluator<T, R>,
    opts: InheritedOptions<T, R> = {}
  ): Attribute<T, R> {
    return reifyInheritedAttribute<T, R>(this.root, this.list, f, opts);
  }
  syn<R>(
    f: SyntheticAttributeEvaluator<T, R>,
    opts: SyntheticOptions<T, R> = {}
  ): Attribute<T, R> {
    return reifySyntheticAttribute<T, R>(this.root, this.list, f, opts);
  }
  // There are no options here mainly because memoization probably isn't useful.
  der<R>(f: DerivedEvaluator<T, R>): Attribute<T, R> {
    // TODO: As attribute gets expanded, more will probably be needed here.
    return (x: T) => f(x);
  }
}

/**
 * Walk the specified tree starting at node `cur`
 * @param cur Node we start walking from
 * @param tree The tree we are walking
 * @param f Function to evaluate at each node
 */
export function walkTree<T>(cur: T, list: ListChildren<T>, f: (x: T) => void) {
  f(cur);
  const children = list(cur);
  if (Array.isArray(children)) {
    /** If this is an indexed tree... */
    for (const child of children) {
      walkTree(child, list, f);
    }
  } else {
    /** If this tree has named children */
    for (const child of Object.entries(children)) {
      walkTree(child[1], list, f);
    }
  }
}

export function childrenOfNode<T>(list: ListChildren<T>, cur: T): Array<T> {
  const children = list(cur);
  return Array.isArray(children)
    ? children
    : Object.entries(children).map((x) => x[1]);
}
