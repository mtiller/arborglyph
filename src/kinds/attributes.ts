import { Arbor } from "../arbor";

/** A scalar function is just a function that takes a single argument that returns a value. */
export type ScalarFunction<T, R> = (x: T) => R;

export interface Attribute<T, R> extends ScalarFunction<T, R> {
  // reroot(): void;
}

export interface AttributeDefinition<T extends object, R> {
  attach(tree: Arbor<T>): Attribute<T, R>;
}
