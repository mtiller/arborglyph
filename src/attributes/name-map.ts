import { SyntheticAttributeEvaluator } from "..";
import { ScalarFunction } from "../kinds/attributes";

export function symbolTableEvaluator<T>(
  namer: ScalarFunction<T, string>
): SyntheticAttributeEvaluator<T, Map<string, T>> {
  return ({ node, children }) =>
    children.reduce((p, c) => {
      c.attr.forEach((v, k) => p.set(k, v));
      return p;
    }, new Map<string, T>([[namer(node), node]]));
}

// TODO: Conditional name map
