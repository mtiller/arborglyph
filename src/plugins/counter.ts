import { Arbor } from "../arbor";
import { Attribute } from "../kinds/attributes";
import { AttributeDefinition } from "../kinds/definitions";
import { ArborPlugin } from "../plugin";
import { CounterStatistics } from "./debug";

export class CounterPlugin<T extends object>
  implements ArborPlugin<T>, CounterStatistics<T>
{
  protected invocationMap: Map<Function, number> = new Map();
  protected nodeInvocationMap: Map<Function, WeakMap<T, number>> = new Map();
  protected evaluationMap: Map<Attribute<T, any>, number> = new Map();
  protected nodeEvaluationMap: Map<Attribute<T, any>, WeakMap<T, number>> =
    new Map();
  connect(arbor: Arbor<T>) {
    arbor.monitor.on("invocation", (a, n) => {
      this.recordInvocation(a, n);
    });
    arbor.monitor.on("evaluation", (d, n) => {
      this.recordEvaluation(d, n);
    });
  }
  protected recordInvocation(d: AttributeDefinition<T, any>, n: T): void {
    this.invocationMap.set(d.f, (this.invocationMap.get(d.f) ?? 0) + 1);
    const tallies = this.nodeInvocationMap.get(d.f) ?? new WeakMap();
    tallies.set(n, (tallies.get(n) ?? 0) + 1);
    this.nodeInvocationMap.set(d.f, tallies);
  }
  protected recordEvaluation(a: Attribute<T, any>, n: T): void {
    this.evaluationMap.set(a, (this.evaluationMap.get(a) ?? 0) + 1);
    const tallies = this.nodeEvaluationMap.get(a) ?? new WeakMap();
    tallies.set(n, (tallies.get(n) ?? 0) + 1);
    this.nodeEvaluationMap.set(a, tallies);
  }
  invocations(d: AttributeDefinition<T, any>, n?: T): number {
    if (n === undefined) return this.invocationMap.get(d.f) ?? 0;
    const tallies = this.nodeInvocationMap.get(d.f) ?? new WeakMap();
    return tallies.get(n) ?? 0;
  }
  evaluations(d: Attribute<T, any>, n?: T): number {
    if (n === undefined) return this.evaluationMap.get(d) ?? 0;
    const tallies = this.nodeEvaluationMap.get(d) ?? new WeakMap();
    return tallies.get(n) ?? 0;
  }
}
