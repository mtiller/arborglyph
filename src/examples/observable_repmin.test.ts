/** An implementation of the repmin example */

import {
  findChild,
  fork,
  indexedBinaryChildren,
  leaf,
  LeafNode,
} from "../testing";
import { sampleTree1, SimpleBinaryTree } from "../testing";
import { Arbor } from "../arbor";
import { configure, observable } from "mobx";
import { lruPlugin } from "../plugins/lru";
import { CounterPlugin } from "../plugins/counter";
import { subTable } from "../attributes/name-map";
import { evalPath } from "../attributes/path";
import { DebugPlugin } from "../plugins/debug";
import { evalGlobmin, evalMin, evalRepmin } from "./repmin_attrs";

describe("Run some repmin test cases on mutable trees", () => {
  // Being a bit sloppy with MobX here.
  configure({ enforceActions: "never" });

  it("should handle a basic repmin", () => {
    const logger = new DebugPlugin<SimpleBinaryTree>();
    logger.events.push("Pre-Tree");
    expect(logger.events).toMatchSnapshot();
    const stats = new CounterPlugin();
    const otree = observable(sampleTree1);
    // const otree = sampleTree1;
    const tree = new Arbor(otree, indexedBinaryChildren, {
      plugins: [stats, logger],
      immutable: false,
    });
    logger.events.push("Post-Tree");
    expect(logger.events).toMatchSnapshot();
    const { globmin, repmin, path } = tree.attach(repminCluster);

    // expect(path(tree.root)).toEqual("root");
    // const l = findChild(tree.root, ["left"]);
    // expect(path(l)).toEqual("root/left");

    logger.stringifyNode = path;
    logger.events.push("Post-Attach");
    expect(logger.events).toMatchSnapshot();

    expect(globmin(tree.root)).toEqual(1);
    expect(globmin(tree.root)).toEqual(1);
    logger.events.push("Root Globmin");
    expect(logger.events).toMatchSnapshot();

    expect(repmin(tree.root)).toEqual(repminResult(1));
    logger.events.push("Post root repmin");
    expect(logger.events).toMatchSnapshot();
    expect(stats.invocations(evalMin)).toEqual(53);
    expect(repmin(tree.root)).toEqual(repminResult(1));
    expect(stats.invocations(evalMin)).toEqual(53);
  });

  it("should handle a basic repmin with weakmap caching", () => {
    const stats = new CounterPlugin();
    const otree = observable(sampleTree1);
    const tree = new Arbor(otree, indexedBinaryChildren, {
      plugins: [stats],
      reification: { memoize: true },
      immutable: false,
    });
    const { repmin } = tree.attach(repminCluster);

    expect(repmin(tree.root)).toEqual(repminResult(1));
    expect(stats.invocations(evalMin)).toEqual(15);
    expect(repmin(tree.root)).toEqual(repminResult(1));
    expect(stats.invocations(evalMin)).toEqual(15);
  });

  it("should handle a basic repmin with small caching", () => {
    const stats = new CounterPlugin();
    const otree = observable(sampleTree1);
    const tree = new Arbor(otree, indexedBinaryChildren, {
      plugins: [stats, lruPlugin({ max: 5 })],
      immutable: false,
    });
    const { repmin } = tree.attach(repminCluster);

    expect(repmin(tree.root)).toEqual(repminResult(1));
    expect(stats.invocations(evalMin)).toEqual(42); // Better, but not as good as with weakmap
    expect(repmin(tree.root)).toEqual(repminResult(1));
    expect(stats.invocations(evalMin)).toEqual(42);
  });

  it("should handle a basic repmin with large caching", () => {
    const stats = new CounterPlugin();
    const otree = observable(sampleTree1);
    const tree = new Arbor(otree, indexedBinaryChildren, {
      plugins: [stats, lruPlugin({ max: 15 })],
      immutable: false,
    });
    const { repmin } = tree.attach(repminCluster);

    expect(repmin(tree.root)).toEqual(repminResult(1));
    expect(stats.invocations(evalMin)).toEqual(15); // Matches weak map
    expect(repmin(tree.root)).toEqual(repminResult(1));
    expect(stats.invocations(evalMin)).toEqual(15);
  });
});

export function repminCluster(tree: Arbor<SimpleBinaryTree>) {
  const min = tree.add(evalMin);
  const globmin = tree.add(evalGlobmin(min));
  const repminAttr = evalRepmin(globmin);
  const repmin = tree.add(repminAttr, {});
  const path = tree.add(evalPath({ eager: false }));
  const subtable = tree.add(
    subTable(path, (x): x is LeafNode => x.type === "leaf")
  );
  // Invoking an attribute here causes all kinds of premature evaluations
  // which interfere with my ability to test lazy evaluation.
  //const table = subtable(tree.root);

  return { subtable, min, globmin, repmin, path, repminAttr };
}

export function repminResult(x: number) {
  return fork(
    fork(leaf(x), fork(fork(leaf(x), leaf(x)), leaf(x))),
    fork(leaf(x), fork(leaf(x), fork(leaf(x), leaf(x))))
  );
}
