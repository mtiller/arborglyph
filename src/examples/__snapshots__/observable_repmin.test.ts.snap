// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Run some repmin test cases on mutable trees should handle a basic repmin 1`] = `
Array [
  "Pre-Tree",
]
`;

exports[`Run some repmin test cases on mutable trees should handle a basic repmin 2`] = `
Array [
  "Pre-Tree",
  "Post-Tree",
]
`;

exports[`Run some repmin test cases on mutable trees should handle a basic repmin 3`] = `
Array [
  "Pre-Tree",
  "Post-Tree",
  "Post-Attach",
]
`;

exports[`Run some repmin test cases on mutable trees should handle a basic repmin 4`] = `
Array [
  "Pre-Tree",
  "Post-Tree",
  "Post-Attach",
  "path(root) => \\"root\\"",
  "path(root/left) => \\"root/left\\"",
  "path(root/left/left) => \\"root/left/left\\"",
  "min(root/left/left) => 7",
  "path(root/left/right) => \\"root/left/right\\"",
  "path(root/left/right/left) => \\"root/left/right/left\\"",
  "path(root/left/right/left/left) => \\"root/left/right/left/left\\"",
  "min(root/left/right/left/left) => 5",
  "path(root/left/right/left/right) => \\"root/left/right/left/right\\"",
  "min(root/left/right/left/right) => 9",
  "min(root/left/right/left) => 5",
  "path(root/left/right/right) => \\"root/left/right/right\\"",
  "min(root/left/right/right) => 3",
  "min(root/left/right) => 3",
  "min(root/left) => 3",
  "path(root/right) => \\"root/right\\"",
  "path(root/right/left) => \\"root/right/left\\"",
  "min(root/right/left) => 4",
  "path(root/right/right) => \\"root/right/right\\"",
  "path(root/right/right/left) => \\"root/right/right/left\\"",
  "min(root/right/right/left) => 1",
  "path(root/right/right/right) => \\"root/right/right/right\\"",
  "path(root/right/right/right/left) => \\"root/right/right/right/left\\"",
  "min(root/right/right/right/left) => 6",
  "path(root/right/right/right/right) => \\"root/right/right/right/right\\"",
  "min(root/right/right/right/right) => 8",
  "min(root/right/right/right) => 6",
  "min(root/right/right) => 1",
  "min(root/right) => 1",
  "min(root) => 1",
  "global min(root) => 1",
  "Root Globmin",
]
`;

exports[`Run some repmin test cases on mutable trees should handle a basic repmin 5`] = `
Array [
  "Pre-Tree",
  "Post-Tree",
  "Post-Attach",
  "path(root) => \\"root\\"",
  "path(root/left) => \\"root/left\\"",
  "path(root/left/left) => \\"root/left/left\\"",
  "min(root/left/left) => 7",
  "path(root/left/right) => \\"root/left/right\\"",
  "path(root/left/right/left) => \\"root/left/right/left\\"",
  "path(root/left/right/left/left) => \\"root/left/right/left/left\\"",
  "min(root/left/right/left/left) => 5",
  "path(root/left/right/left/right) => \\"root/left/right/left/right\\"",
  "min(root/left/right/left/right) => 9",
  "min(root/left/right/left) => 5",
  "path(root/left/right/right) => \\"root/left/right/right\\"",
  "min(root/left/right/right) => 3",
  "min(root/left/right) => 3",
  "min(root/left) => 3",
  "path(root/right) => \\"root/right\\"",
  "path(root/right/left) => \\"root/right/left\\"",
  "min(root/right/left) => 4",
  "path(root/right/right) => \\"root/right/right\\"",
  "path(root/right/right/left) => \\"root/right/right/left\\"",
  "min(root/right/right/left) => 1",
  "path(root/right/right/right) => \\"root/right/right/right\\"",
  "path(root/right/right/right/left) => \\"root/right/right/right/left\\"",
  "min(root/right/right/right/left) => 6",
  "path(root/right/right/right/right) => \\"root/right/right/right/right\\"",
  "min(root/right/right/right/right) => 8",
  "min(root/right/right/right) => 6",
  "min(root/right/right) => 1",
  "min(root/right) => 1",
  "min(root) => 1",
  "global min(root) => 1",
  "Root Globmin",
  "min(root/left/left) => 7",
  "min(root/left/right/left/left) => 5",
  "min(root/left/right/left/right) => 9",
  "min(root/left/right/left) => 5",
  "min(root/left/right/right) => 3",
  "min(root/left/right) => 3",
  "min(root/left) => 3",
  "global min(root/left) => 1",
  "min(root/left/left) => 7",
  "global min(root/left/left) => 1",
  "repmin(root/left/left) => {\\"type\\":\\"leaf\\",\\"value\\":1}",
  "min(root/left/right/left/left) => 5",
  "min(root/left/right/left/right) => 9",
  "min(root/left/right/left) => 5",
  "min(root/left/right/right) => 3",
  "min(root/left/right) => 3",
  "global min(root/left/right) => 1",
  "min(root/left/right/left/left) => 5",
  "min(root/left/right/left/right) => 9",
  "min(root/left/right/left) => 5",
  "global min(root/left/right/left) => 1",
  "min(root/left/right/left/left) => 5",
  "global min(root/left/right/left/left) => 1",
  "repmin(root/left/right/left/left) => {\\"type\\":\\"leaf\\",\\"value\\":1}",
  "min(root/left/right/left/right) => 9",
  "global min(root/left/right/left/right) => 1",
  "repmin(root/left/right/left/right) => {\\"type\\":\\"leaf\\",\\"value\\":1}",
  "repmin(root/left/right/left) => {\\"type\\":\\"fork\\",\\"left\\":{\\"type\\":\\"leaf\\",\\"value\\":1},\\"right\\":{\\"type\\":\\"leaf\\",\\"value\\":1}}",
  "min(root/left/right/right) => 3",
  "global min(root/left/right/right) => 1",
  "repmin(root/left/right/right) => {\\"type\\":\\"leaf\\",\\"value\\":1}",
  "repmin(root/left/right) => {\\"type\\":\\"fork\\",\\"left\\":{\\"type\\":\\"fork\\",\\"left\\":{\\"type\\":\\"leaf\\",\\"value\\":1},\\"right\\":{\\"type\\":\\"leaf\\",\\"value\\":1}},\\"right\\":{\\"type\\":\\"leaf\\",\\"value\\":1}}",
  "repmin(root/left) => {\\"type\\":\\"fork\\",\\"left\\":{\\"type\\":\\"leaf\\",\\"value\\":1},\\"right\\":{\\"type\\":\\"fork\\",\\"left\\":{\\"type\\":\\"fork\\",\\"left\\":{\\"type\\":\\"leaf\\",\\"value\\":1},\\"right\\":{\\"type\\":\\"leaf\\",\\"value\\":1}},\\"right\\":{\\"type\\":\\"leaf\\",\\"value\\":1}}}",
  "min(root/right/left) => 4",
  "min(root/right/right/left) => 1",
  "min(root/right/right/right/left) => 6",
  "min(root/right/right/right/right) => 8",
  "min(root/right/right/right) => 6",
  "min(root/right/right) => 1",
  "min(root/right) => 1",
  "global min(root/right) => 1",
  "min(root/right/left) => 4",
  "global min(root/right/left) => 1",
  "repmin(root/right/left) => {\\"type\\":\\"leaf\\",\\"value\\":1}",
  "min(root/right/right/left) => 1",
  "min(root/right/right/right/left) => 6",
  "min(root/right/right/right/right) => 8",
  "min(root/right/right/right) => 6",
  "min(root/right/right) => 1",
  "global min(root/right/right) => 1",
  "min(root/right/right/left) => 1",
  "global min(root/right/right/left) => 1",
  "repmin(root/right/right/left) => {\\"type\\":\\"leaf\\",\\"value\\":1}",
  "min(root/right/right/right/left) => 6",
  "min(root/right/right/right/right) => 8",
  "min(root/right/right/right) => 6",
  "global min(root/right/right/right) => 1",
  "min(root/right/right/right/left) => 6",
  "global min(root/right/right/right/left) => 1",
  "repmin(root/right/right/right/left) => {\\"type\\":\\"leaf\\",\\"value\\":1}",
  "min(root/right/right/right/right) => 8",
  "global min(root/right/right/right/right) => 1",
  "repmin(root/right/right/right/right) => {\\"type\\":\\"leaf\\",\\"value\\":1}",
  "repmin(root/right/right/right) => {\\"type\\":\\"fork\\",\\"left\\":{\\"type\\":\\"leaf\\",\\"value\\":1},\\"right\\":{\\"type\\":\\"leaf\\",\\"value\\":1}}",
  "repmin(root/right/right) => {\\"type\\":\\"fork\\",\\"left\\":{\\"type\\":\\"leaf\\",\\"value\\":1},\\"right\\":{\\"type\\":\\"fork\\",\\"left\\":{\\"type\\":\\"leaf\\",\\"value\\":1},\\"right\\":{\\"type\\":\\"leaf\\",\\"value\\":1}}}",
  "repmin(root/right) => {\\"type\\":\\"fork\\",\\"left\\":{\\"type\\":\\"leaf\\",\\"value\\":1},\\"right\\":{\\"type\\":\\"fork\\",\\"left\\":{\\"type\\":\\"leaf\\",\\"value\\":1},\\"right\\":{\\"type\\":\\"fork\\",\\"left\\":{\\"type\\":\\"leaf\\",\\"value\\":1},\\"right\\":{\\"type\\":\\"leaf\\",\\"value\\":1}}}}",
  "repmin(root) => {\\"type\\":\\"fork\\",\\"left\\":{\\"type\\":\\"fork\\",\\"left\\":{\\"type\\":\\"leaf\\",\\"value\\":1},\\"right\\":{\\"type\\":\\"fork\\",\\"left\\":{\\"type\\":\\"fork\\",\\"left\\":{\\"type\\":\\"leaf\\",\\"value\\":1},\\"right\\":{\\"type\\":\\"leaf\\",\\"value\\":1}},\\"right\\":{\\"type\\":\\"leaf\\",\\"value\\":1}}},\\"right\\":{\\"type\\":\\"fork\\",\\"left\\":{\\"type\\":\\"leaf\\",\\"value\\":1},\\"right\\":{\\"type\\":\\"fork\\",\\"left\\":{\\"type\\":\\"leaf\\",\\"value\\":1},\\"right\\":{\\"type\\":\\"fork\\",\\"left\\":{\\"type\\":\\"leaf\\",\\"value\\":1},\\"right\\":{\\"type\\":\\"leaf\\",\\"value\\":1}}}}}",
  "Post root repmin",
]
`;
